plugins/parsers/carbon2/parser.go

```diff
@@ -0,0 +1,183 @@
+package carbon2
+
+import (
+	"bufio"
+	"bytes"
+	"errors"
+	"fmt"
+	"io"
+	"strconv"
+	"time"
+
+	"github.com/influxdata/telegraf"
+	"github.com/influxdata/telegraf/metric"
+)
+
+const (
+	tagNameField  = "field"
+	tagNameMetric = "metric"
+)
+
+type Parser struct {
+}
+
+// Parse takes a byte buffer separated by newlines
+// ie, `cpu.usage.idle 90\ncpu.usage.busy 10` and parses it into telegraf metrics.
+//
+// Must be thread-safe.
+func (p Parser) Parse(buf []byte) ([]telegraf.Metric, error) {
+	// parse even if the buffer begins with a newline
+	buf = bytes.TrimPrefix(buf, []byte("\n"))
+
+	var (
+		buffer  = bytes.NewBuffer(buf)
+		reader  = bufio.NewReader(buffer)
+		metrics []telegraf.Metric
+	)
+
+	for {
+		line, err := reader.ReadBytes('\n')
+		if err != nil && err != io.EOF {
+			return nil, err
+		}
+		if err == io.EOF && len(line) == 0 {
+			break
+		}
+
+		m, err := parseLine(line)
+		if err != nil {
+			return nil, fmt.Errorf("failed to parse line: %s, err: %w", line, err)
+		}
+
+		metrics = append(metrics, m)
+	}
+
+	return metrics, nil
+}
+
+func parseLine(line []byte) (telegraf.Metric, error) {
+	var (
+		buf      = bytes.NewBuffer(line)
+		name     string
+		tim      time.Time
+		tags     = make(map[string]string)
+		fields   = make(map[string]interface{})
+		gotValue bool
+	)
+
+	for {
+		bb, err := buf.ReadBytes(' ')
+		if err != nil && err != io.EOF {
+			return nil, err
+		}
+
+		if bytes.HasPrefix(bb, []byte(" ")) {
+			continue
+		}
+
+		idx := bytes.IndexByte(bb, '=')
+		if idx == -1 {
+			if !gotValue {
+				// It's a value so parse it
+				v, err := parseBytesForValue(bb)
+				if err != nil {
+					return nil, err
+				}
+
+				// Note: this works around the fact that carbon2 serializer can
+				// either:
+				// * stitch together using '_' telegraf's metric Name and field
+				//   name (taken from 'field' tag) and take that to be used
+				//   as metric name, e.g. metric=memory_available
+				// * use a separate 'field' tag, e.g. metric=memory field=available
+				//
+				// Hence parsing and serializing a metric would yield a different
+				// result then ingested. Because of that reason we
+				fields[""] = v
+				gotValue = true
+				continue
+			} else {
+				// It's a timestamp so parse it
+				t, err := parseBytesForTimestamp(bb)
+				if err != nil {
+					return nil, err
+				}
+
+				tim = t
+				break
+			}
+		}
+
+		tag, value, err := getTag(bb, idx)
+		if err != nil {
+			return nil, err
+		}
+
+		if bytes.Compare(tag, []byte(tagNameMetric)) == 0 {
+			// If it's a 'metric' tag then set it as metric's name
+			name = string(value)
+		} else if bytes.Compare(tag, []byte(tagNameField)) == 0 {
+			// If it's a 'field' tag then set it as 'field'
+			tags["field"] = string(value)
+		} else {
+			tags[string(tag)] = string(value)
+		}
```

Comment:
This is same code for `field` and non `field`

Reply:
Done